### PR TITLE
Feature/cdap 14770 migrate NamespaceMDS

### PR DESCRIPTION
**Pull request by jayfjin**
Migrate namespace MDS to the new structuredtable SPI. 

build ongoing: https://builds.cask.co/browse/CDAP-DUT6750

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java [hunk 1]
```diff
@@ -79,14 +81,15 @@
     new GsonBuilder().registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())).create();
 
   private final MetadataStore metadataStore;
-  private final NamespaceMDS namespaceMDS;
+  private final DefaultNamespaceStore defaultNamespaceStore;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/profile/ProfileMetadataMessageProcessor.java]: can this be a NamespaceStore? Or do you rely on anything that is not in the interface?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java [hunk 2]
```diff
@@ -89,28 +90,22 @@
 /**
  * Tests for {@link DefaultStore}.
  */
-public class DefaultStoreTest {
+public abstract class DefaultStoreTest {
+  protected static DefaultStore store;
+  protected static DefaultNamespaceStore nsStore;
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java:95]: why is a namespace store and a namespace admin needed? Are they the same thing?

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/NamespaceMDSTable.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset for namespace metadata. It does not wrap its operations in a transaction.
+ * It is up to the caller to decide what operations belong in a transaction.
+ */
+public final class NamespaceMDSTable {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/NamespaceMDSTable.java]: could it be named just "NamespaceTable"?

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/DefaultNamespaceStore.java [hunk 4]
```diff
@@ -45,33 +33,23 @@
...
-    );
+  public DefaultNamespaceStore(TransactionRunner transactionRunner) {
+    this.transactionRunner = transactionRunner;
   }
 
-  private NamespaceMDS getNamespaceMDS(DatasetContext datasetContext) throws IOException, DatasetManagementException {
-    Table table = DatasetsUtil.getOrCreateDataset(datasetContext, dsFramework, APP_META_INSTANCE_ID,
-                                                  Table.class.getName(), DatasetProperties.EMPTY);
-    return new NamespaceMDS(table);
+  private NamespaceMDSTable getNamespaceMDS(StructuredTableContext context) throws NotFoundException {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/DefaultNamespaceStore.java]: name this method getNamespaceTable()?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/SqlDefaultStoreTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+  @BeforeClass
+  public static void beforeClass() throws IOException, AlreadyExistsException {
+    Injector injector = AppFabricTestHelper.getInjector();
+    // TODO(CDAP-14770): change this when migrating DefaultStore
+    store = injector.getInstance(DefaultStore.class);
+
+    EmbeddedPostgres pg = EmbeddedPostgres.start();
+    DataSource dataSource = pg.getPostgresDatabase();
+    StructuredTableAdmin structuredTableAdmin = new PostgresSqlStructuredTableAdmin(dataSource);
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/SqlDefaultStoreTest.java:55]: I assume that this opens a JDBC connection somewhere? Does that mean we also need to close that when the test shuts down? And more generally, everywhere we use structured tables?

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 6]
```diff
@@ -44,6 +43,31 @@ public static void createAllTables(StructuredTableAdmin tableAdmin) throws IOExc
...
+    if (tableAdmin.getSpecification(NamespaceStore.NAMESPACE_MDS_TABLE) == null) {
+      NamespaceStore.createTables(tableAdmin);
+    }
+  }
+
+  /**
+   * Namespace store schema
+   */
+  public static final class NamespaceStore {
+    public static final StructuredTableId NAMESPACE_MDS_TABLE = new StructuredTableId("namespace_mds");
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: is the _MDS_ needed?

**Review by anew — COMMENTED** [review]
Overall looks good, please see my comments

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/SqlDefaultStoreTest.java (hunk 5 above)
Comment (jayfjin) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/SqlDefaultStoreTest.java:55]: Good point, changed this and the sql artifact store test.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java (hunk 2 above)
Comment (jayfjin) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java:95]: We cannot delete the default namespace in namespace admin.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/SqlArtifactStoreTest.java [hunk 7]
```diff
@@ -55,4 +60,9 @@ public static void setup() throws Exception {
                                       transactionRunner);
     StoreDefinition.ArtifactStore.createTables(structuredTableAdmin);
   }
+
+  @AfterClass
+  public static void afterClass() throws IOException {
+    pg.close();
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/SqlArtifactStoreTest.java:66]: I assume this stops the embedded postgres?

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 8]
```diff
@@ -52,21 +52,21 @@ public static void createAllTables(StructuredTableAdmin tableAdmin) throws IOExc
    * Namespace store schema
    */
   public static final class NamespaceStore {
-    public static final StructuredTableId NAMESPACE_MDS_TABLE = new StructuredTableId("namespace_mds");
+    public static final StructuredTableId NAMESPACE_TABLE = new StructuredTableId("namespace_table");
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: I feel that this would be properly named "namespaces". What do you think?

**Review by anew — APPROVED** [review]
Two minor comments, please address and merge without further review. Also, please make sure to update the copyright for al changed files to 2019 :)

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/SqlArtifactStoreTest.java (hunk 7 above)
Comment (jayfjin) [reply cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/artifact/SqlArtifactStoreTest.java:66]: correct

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java (hunk 8 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: sg, thanks for the review!

**Pull request merged**